### PR TITLE
Implemented unique_ptr

### PR DESCRIPTION
**Pull request by J-Cieplinski**

### CODE REVIEW

**Review by jakjan95 — APPROVED** [review]
Good Code :) Try to add delete statement for copy assignment  and it will be perfect :)

File: .github/workflows/unique_ptr.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+name: unique_ptr Homework
+
+on: [push, pull_request]
+
+jobs:
+  unique_ptr:
```
Comment (ziobron) [inline .github/workflows/unique_ptr.yaml:6]: an additional job with valgrind would prove that there are no memory leaks

File: homework/unique_ptr/test.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+struct TestingPtr : ::testing::Test {
+    TestingPtr()
+        : ptr(new std::string("Ala ma kota")) {
+    }
+
+    cs::unique_ptr<std::string> ptr;
+};
+
+TEST_F(TestingPtr, ShouldProperlyDereference) {
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:15]: properly is a forbidden word in test names ;)

File: homework/unique_ptr/test.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    std::string testString{"Ala ma kota"};
+
+    ASSERT_EQ(testString, *ptr);
+}
+
+TEST_F(TestingPtr, ShouldProperlyReturnPointerViaGet) {
+    auto testPtr = ptr.get();
+    ASSERT_EQ(testPtr, ptr.get());
+}
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:24]: this test is testing something else. It checks if you can use `get()` twice with the same result. 
It shoudn't work for `release()` function.

File: homework/unique_ptr/test.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+TEST_F(TestingPtr, ShouldProperlyReset) {
+    std::string valueToResetTo{"Kota ma Ale"};
+    ptr.reset(new std::string(valueToResetTo));
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:28]: what about reset without an argument? :)
`ptr.reset()`

File: homework/unique_ptr/test.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+    ASSERT_EQ(valueToResetTo, *ptr);
+}
+
+TEST_F(TestingPtr, ShouldProperlyMovePointer) {
+    cs::unique_ptr<std::string> testPointer(std::move(ptr));
+
+    ASSERT_EQ(testString, *testPointer);
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:36]: you should also check what happens to ptr after std::move

File: homework/unique_ptr/test.cpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+TEST_F(TestingPtr, ShouldUseMoveAssignmentOperator) {
+    cs::unique_ptr<std::string> testPointer(nullptr);
+    testPointer = std::move(ptr);
+
+    ASSERT_EQ(testString, *testPointer);
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:43]: here as well, check what happens to ptr

File: homework/unique_ptr/test.cpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+TEST_F(TestingPtr, ShouldProperlyMovePointer) {
+    cs::unique_ptr<std::string> testPointer(std::move(ptr));
+
+    ASSERT_EQ(testString, *testPointer);
+}
+
+TEST_F(TestingPtr, ShouldUseMoveAssignmentOperator) {
+    cs::unique_ptr<std::string> testPointer(nullptr);
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:40]: can I initialize cs::unique_ptr with default constructor here?
```suggestion
    cs::unique_ptr<std::string> testPointer{};
```

File: homework/unique_ptr/test.cpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    testPointer = std::move(ptr);
+
+    ASSERT_EQ(testString, *testPointer);
+}
+
+TEST_F(TestingPtr, MovedPointerShouldBeNullptr) {
+    cs::unique_ptr<std::string> testPointer(std::move(ptr));
+
+    ASSERT_EQ(nullptr, ptr.get());
```
Comment (ziobron) [inline homework/unique_ptr/test.cpp:49]: ok, here you have checked what happened to the source pointer. It should be integrated into one test in my opinion.

File: homework/unique_ptr/unique_ptr.hpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+};
+
+template <typename T>
+unique_ptr<T>::unique_ptr(T* ptr)
+    : ptr_(ptr) {
+}
+
+template <typename T>
+unique_ptr<T>::unique_ptr(unique_ptr&& previousOwner) {
+    ptr_ = previousOwner.release();
```
Comment (ziobron) [inline homework/unique_ptr/unique_ptr.hpp:36]: what will happen it the target pointer already manages an object? I guess it's a memory leak. Please add a proper test case.

File: homework/unique_ptr/unique_ptr.hpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        ptr_ = previousOwner.ptr_;
+        previousOwner.ptr_ = nullptr;
+    }
+
+    return *this;
+}
+
+template <typename T>
+unique_ptr<T>::~unique_ptr() {
+    if (ptr_) {
```
Comment (ziobron) [inline homework/unique_ptr/unique_ptr.hpp:51]: `if` not needed. calling delete on nullptr is safe.

File: homework/unique_ptr/unique_ptr.hpp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    return tempPtr;
+}
+
+template <typename T>
+void unique_ptr<T>::reset(T* newPtr) {
+    T* oldPtr = ptr_;
+    ptr_ = newPtr;
+    if (oldPtr) {
+        delete oldPtr;
+    }
```
Comment (ziobron) [inline homework/unique_ptr/unique_ptr.hpp:88]:
```suggestion
    delete ptr_;
    ptr_ = newPtr;
```

File: homework/unique_ptr/unique_ptr.hpp [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+public:
+    unique_ptr(T* ptr);
+    unique_ptr() = delete;
+    unique_ptr(unique_ptr&& previousOwner);
+    unique_ptr(const unique_ptr&) = delete;
+    ~unique_ptr();
+
+    T* release();
+    const T* get() const;
+    void reset(T* newPtr);
```
Comment (ziobron) [inline homework/unique_ptr/unique_ptr.hpp:18]:
```suggestion
    void reset(T* newPtr = nullptr);
```
Default parameter would be useful

**Review by ziobron — CHANGES_REQUESTED** [review]
Probably there may be a memory leak in one case. Please check it.

File: homework/unique_ptr/test.cpp (hunk 5 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/test.cpp:36]: Already done in different test `MovedPointerShouldBeNullptr`

File: homework/unique_ptr/test.cpp (hunk 6 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/test.cpp:43]: Added `MovedAssignedPointerShouldBeNullptr`

File: homework/unique_ptr/test.cpp (hunk 8 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/test.cpp:49]: Oh well, here I go reverting changes :D

File: homework/unique_ptr/unique_ptr.hpp (hunk 9 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/unique_ptr.hpp:36]: Well, it shouldn't, it's a constructor :D

File: homework/unique_ptr/test.cpp (hunk 7 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/test.cpp:40]: After commit, yes

File: .github/workflows/unique_ptr.yaml (hunk 1 above)
Comment (J-Cieplinski) [reply .github/workflows/unique_ptr.yaml:6]: done

File: homework/unique_ptr/test.cpp (hunk 3 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/test.cpp]: Done, i think?

File: homework/unique_ptr/test.cpp (hunk 2 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/test.cpp]: Done

File: homework/unique_ptr/test.cpp (hunk 4 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/test.cpp:29]: Done

File: homework/unique_ptr/test.cpp (hunk 5 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/test.cpp:44]: merged tests

File: homework/unique_ptr/test.cpp (hunk 6 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/test.cpp:51]: Done

File: homework/unique_ptr/unique_ptr.hpp (hunk 12 above)
Comment (J-Cieplinski) [reply homework/unique_ptr/unique_ptr.hpp]: I've done it differently now

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
{ "requirement_ids": [3784, 1119] }
36 XP za implementację unique_ptra (3784)
3 XP za punktualność

**Pull request closed without merging**